1-Draft/RFCNNNN-Azure-IoT-PowerShell-Device-Sdk.md

```diff
@@ -0,0 +1,186 @@
+The barebones service.
+Think of it like an event hub in a way.
+* [Azure IoT Suite](https://docs.microsoft.com/en-us/azure/iot-suite/) -
+A "preconfigured solution" that includes an instance of Azure IoT Hub and other services with everything set up out of the box.
+
+## Motivation
+
+The PowerShell team is investigating IoT scenarios for PowerShell.
+
+This module will be separate to a PowerShell IoT module
+(described in a separate RFC - linked to later),
+but we want to enable scenario that will allow folks to use PowerShell IoT with Azure.
+
+In order to enable that,
+we need a PowerShell module that will wrap the .NET device SDK.
+The SDK caters to C# and we want to give users of PowerShell and PowerShell IoT a "PowerShell-y" experience when interacting with Azure IoT so they can iterate fast with a familiar interface.
+
+## Specification
+
+As mentioned,
+[Work has already been started](https://github.com/tylerl0706/azure-iot-sdk-powershell).
+Here are the cmdlet interfaces we want to ship initially:
+
+### Connecting to Azure IoT Hub
+```powershell
+$ConnectionString = "<your conn string>"
+Connect-AzureIoTDevice -ConnectionString $ConnectionString
+```
+
+### Setting Azure IoT Reported Properties (the device config/metadata, essentially)
+```powershell
+$reportedPropertiesObj = @{
+    MyData = "Hello World"
+}
+Set-IoTDeviceReportedProperties -ReportedProperties $reportedPropertiesObj
+```
+
+### Sending a message from the device to the Azure IoT Hub
+
+```powershell
+Invoke-IoTDeviceEvent -Message "Hello World"
+```
+NOTE: `Connect-AzureIoTDevice` returns a `DeviceClient` object which can optionally be passed in to `Set-IoTDeviceReportedProperties` or `Invoke-IoTDeviceEvent` to specify the device client you want to use.
+If it is not passed in,
+the cmdlets will use the last initialized device client.
+
+### Handling Direct Method invocations (a bit complicated to implement)
+
+The Direct Method portion of the Azure IoT device SDK is asynchronous and,
+in C#,
+calls a delegate (that you define on the device) when the name of the function comes in from the server.
+
+To make it more PowerShell-y,
+we want to be able to run PowerShell functions,
+rather than having the user define C# delegates.
+Because of the nature of async Tasks,
+delegates get run on a separate thread.
+Since PowerShell is single threaded and doesn't quite support Tasks simply,
+we must wrap user defined PowerShell functions in a delegate that can be run on a separate thread that will create a new runspace to execute the PowerShell in.
+
+To define these functions,
+a user can create a PowerShell module like so:
+```powershell
+function Get-AzureIoTManifest {
+    return @{
+        hello = "Get-HelloWorld"
+        echo = "Get-Echo"
+    }
+}
+
+function Get-HelloWorld ($Request) {
+    return @{
+        Hello = "World"
+    }
+}
+
+function Get-Echo ($Request) {
+    return @{
+        Data = $Request
+    }
+}
+
+Export-ModuleMember -Function *-*
```

Comment:
Best practice is to have a module manifest which explicitly exports functions by name rather than using wildcard.  This improves performance of cmdlet discovery.  When wildcards are used, PowerShell has to parse all the files to determine what is exported.